### ISSUE TITLE
ansible-lint>=5 and molecule>=3.3 require symlinks in current version to work (previous versions worked fine)

### ISSUE THREAD

**Issue by dlevene1** [ansible-community/ansible-lint#1555]
# Issue Type

- Bug report

# Molecule and Ansible details

```
 molecule --version && ansible --version
molecule 3.3.1 using python 3.6
    ansible:2.10.3
    delegated:3.3.1 from molecule
    docker:0.3.3 from molecule_docker
ansible 2.10.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/.local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Apr  2 2020, 13:34:55) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

Molecule installation method (one of):

- pip

Ansible installation method (one of):

- pip

Detail any linters or test runners used:

# Desired Behaviour

Previous versions of molecule worked fine. Not sure which version was running, sorry, but when running standard molecule tests things were ok, now i'm hitting an issue where I can't run them due to it wanting to create symlinks (which are not allowed in my setup)

When I copy the role to a spot where symlinks are allowed it works.

`INFO     Using .cache/roles/role_name symlink to current repository in order to enable Ansible to find the role using its expected full name.`

I would like to be able to run this without creating a symlink.

While I can review making future changes and change my way of working, it did work in the past, hence raising this bug report.

# Actual Behaviour

Please give some details of what is actually happening.
```
$ molecule --debug lint
DEBUG    Validating schema /git_repos/myrolename/molecule/default/molecule.yml.
INFO     default scenario test matrix: dependency, lint
INFO     Performing prerun...
Traceback (most recent call last):
  File "/usr/local/bin/molecule", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/molecule/command/lint.py", line 120, in lint
    base.execute_cmdline_scenarios(scenario_name, args, command_args)
  File "/usr/local/lib/python3.6/site-packages/molecule/command/base.py", line 112, in execute_cmdline_scenarios
    prepare_environment()
  File "/home/levenda/.local/lib/python3.6/site-packages/ansiblelint/prerun.py", line 172, in prepare_environment
    _install_galaxy_role()
  File "/home/levenda/.local/lib/python3.6/site-packages/ansiblelint/prerun.py", line 254, in _install_galaxy_role
    link_path.symlink_to(pathlib.Path("../..", target_is_directory=True))
  File "/usr/lib64/python3.6/pathlib.py", line 1327, in symlink_to
    self._accessor.symlink(target, self, target_is_directory)
  File "/usr/lib64/python3.6/pathlib.py", line 434, in symlink
    return os.symlink(str(a), str(b))
PermissionError: [Errno 1] Operation not permitted: '../..' -> '.cache/roles/myrolename'
```

**Comment by ssbarnea** [comment]
Running under Windows is a non supported use-case. Development tools assume the development manching does meet ansible-controller requirements (which is now Linux and a posix filesystem). As these are not met, I will need to close this bug as wontfix.

Still, if you manage to find a workaround and propose PR that address this problem, I will look at it. Still, even with it merged I cannot guarantee you that it will not break in a different place. Symlinks are very popular and useful on linux and we do not avoid making use of them.

**Comment by dlevene1** [comment]
@ssbarnea - understood, I'm running it on linux - but the filesystem doesn't support symlinks. agree they are very useful. 

If I were to look at raising a PR and having the ability to define the temp dir to be created elsewhere .. say /tmp/role_name where would I start looking in the code?

**Comment by ssbarnea** [comment]
@tadeboro Tox also stores .tox inside the repo but pre-commit decided to make use of `~/.cache`. There are pros and cons for each of these. Using `~/.cache` with a hash for each repo would make cleaning it much harder as you need to check which repos are no longer there. Having the in-repo approach has the benefit that contains the project files, once the project dir is removed the temp files are also gone.

What may not be clear to you is that if you copy the files to any destination instead of using symlinks, you lose the ability to receive meaningful error messages about your own code. That is because errors will point to the copied files and not the original source code. You will no longer be able to click a filename in the console and fix it, as this file will be overridden at the next execution. For 3rd party collections and role that is not a problem but these are not the only things that gets copied. If you are inside a collection repo, the entire collection gets installed using symlinks.

I am not sure why @tadeboro dislikes the `project_dir/.cache` so much, but I hope to find a solution.

**Comment by tadeboro** [comment]
I am not talking about removing symlink at all here. I all suggested is to allow users to override the location of the `.cache` directory. This will solve two "issues":

 1. It will give grumpy users like me an easy way of getting temporary files (that I do not need in the first place) into some temporary dir where they will not pollute the project (for example, right now I need to add `.cache` to `.gitignore`to avoid git nagging me about those files and I need to add the same folder to `galaxy.yml` ignore list - and that one does not even work on Ansible 2.9 so the possibility of the garbage finding its way into collection artifact is even higher).
 2. It will unblock the current scenario since directories like `/tmp` probably live on a different file system that does not have the same limitations about the symlinks as the mounted folders have.
